stan/math/fwd/functor/operands_and_partials.hpp

```diff
@@ -2,33 +2,201 @@
+ * @tparam T1 a type with a `.dx()` method.
+ * @param a an edge from `operands_and_partials_impl`
+ */
+template <typename T1>
+inline auto sum_dx(T1& a) {
+  return a.dx();
+}
+
+/**
+ * Accumulate the `.dx()` from each of the edges in `ops_and_partials_impl`.
+ * @tparam T1 a type with a `.dx()` method.
+ * @tparam T2 a type with a `.dx()` method.
+ * @tparam Types a type with a `.dx()` method.
+ * @param a an edge from `operands_and_partials_impl`
+ * @param b an edge from `operands_and_partials_impl`
+ * @param args edges from `operands_and_partials_impl`
+ */
+template <typename T1, typename T2, typename... Types>
+inline auto sum_dx(T1& a, T2& b, Types&... args) {
+  return a.dx() + b.dx() + sum_dx(args...);
+}
+
+template <typename InnerType, typename T>
+class ops_partials_edge<InnerType, T, require_fvar_t<T>> {
+ public:
+  using Op = std::decay_t<T>;
+  using Dx = std::decay_t<InnerType>;
+  Dx partial_{0};
+  broadcast_array<Dx> partials_{partial_};
+
+  explicit ops_partials_edge(const T& op)
+      : partial_(0), partials_(partial_), operands_(op) {}
+
+  explicit ops_partials_edge(
+      const ops_partials_edge<InnerType, T, require_fvar_t<T>>& other)
+      : partial_(other.partial_),
+        partials_(partial_),
+        operands_(other.operands_) {}
+
+  explicit ops_partials_edge(
+      ops_partials_edge<InnerType, T, require_fvar_t<T>>&& other)
+      : partial_(other.partial_),
+        partials_(partial_),
+        operands_(other.operands_) {}
+
+  const Op& operands_;
+
+  inline Dx dx() { return this->partial_ * this->operands_.d_; }
+};
+
+// Vectorized Univariate
+template <typename InnerType, typename T>
+class ops_partials_edge<InnerType, T, require_std_vector_vt<is_fvar, T>> {
+ public:
+  using Op = std::decay_t<T>;
+  using Dx = std::decay_t<InnerType>;
+  using partials_t = Eigen::Matrix<Dx, -1, 1>;
+  partials_t partials_;  // For univariate use-cases
+  broadcast_array<partials_t> partials_vec_{partials_};  // For multivariate
+  explicit ops_partials_edge(const Op& ops)
+      : partials_(partials_t::Zero(ops.size()).eval()), operands_(ops) {}
+
+  explicit ops_partials_edge(
+      const ops_partials_edge<InnerType, T, require_std_vector_vt<is_fvar, T>>&
+          other)
+      : partials_(other.partials_),
+        partials_vec_(partials_),
+        operands_(other.operands_) {}
+
+  explicit ops_partials_edge(
+      ops_partials_edge<InnerType, T, require_std_vector_vt<is_fvar, T>>&&
+          other)
+      : partials_(std::move(other.partials_)),
+        partials_vec_(partials_),
+        operands_(other.operands_) {}
+
+  const Op& operands_;
+  inline Dx dx() {
+    Dx derivative(0);
+    for (size_t i = 0; i < this->operands_.size(); ++i) {
+      derivative += this->partials_[i] * this->operands_[i].d_;
+    }
+    return derivative;
+  }
+};
+
+template <typename Dx, typename ViewElt>
+class ops_partials_edge<Dx, ViewElt, require_eigen_vt<is_fvar, ViewElt>> {
```

Comment:
Would it be possible to have a single template for both the `std::vector<>` and `eigen` types? Not much stands out as being too type-dependent

Reply:
I think there are ways we could shrink it down but I think it might start looking too abstract. The main thing is the matrix types + constructor

```cpp
  explicit ops_partials_edge(const OpT& ops)
      : partials_(partials_t::Zero(ops.rows(), ops.cols())), operands_(ops) {}
```

Reply:
Got it!